[FEAT] 로그인 (인증/인가) 구현

dye0p opened the pull request:
## Issue Number
closed #163

## As-Is
<!-- 문제 상황 정의 -->
- 멘토가 스스로 멘토링을 개설하고, 관리하기 위해 로그인 기능이 필요한 상황
- 멘토링을 마친 멘티가 리뷰를 남기기 위해서 로그인 기능이 필요한 상황

## To-Be
<!-- 변경 사항 -->
- 아이디와 비밀번호를 받아 로그인을 할 수 있다
- 아이디와 비밀번호를 검증한다.
- 로그인시 토큰(JWT)를 발급하며, accessToken과 refreshToken을 쿠키에 담아 응답한다.
- refreshToken은 별도의 테이블이 존재하며 DB에 저장된다.

- **refresh_token 테이블 추가**
<img width="452" height="128" alt="image" src="https://github.com/user-attachments/assets/1ffb8f5d-c202-490a-98d0-14401e715e38" />

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

## (Optional) Additional Description
## 토큰 만료 시간
- accessToken: 15분
- refreshToken: 7일

임의로 정한 만료시간 입니다. 수정이 필요하면 말씀해주세요!

## JWT 선택 이유 (vs 세션)
- 현재 저희 프로젝트 구조 같은 단일 서버 환경에서는 세션 방식이 더 단순합니다.
하지만, 추후 다중 서버나 무중단 배포 환경을 고려하고 있기에, 
세션을 이용하면 세션 공유(Sticky Session, 세션 복제 등)가 필요하여 관리 부담이 증가할 것이라고 판단하였습니다.
- JWT는 서버에 인증 상태를 저장하지 않고(Stateless) 클라이언트가 토큰을 보관하므로 서버 간 상태 동기화 문제가 세션보다 적을 것으로 판단했습니다.

세션도 Redis와 같은 외부 저장소를 사용하면 무중단 배포와 서버 확장에 충분히 대응할 수 있으므로, 
jwt를 반드시 사용해야 하는 것은 아닙니다.
다만, 세션 방식에 비해 세션 저장소 구축, 만료 처리, 서버 간 세션 동기화 등의 관리 포인트가 줄어든다는 점에서,
운영 복잡도를 낮출 수 있다고 판단하여 JWT 방식을 선택하였습니다.

## 토큰 저장 위치
refreshToken과 accessToken 모두 응답 쿠키에 저장했습니다. 이유는 다음과 같습니다.

**refreshToken**
- 만료기간이 긴 토큰으로 만료된 accessToken의 재발급을 위한 토큰 입니다.
- payload에 특별한 정보를 담고있진 않지만, 탈취당하면 해당 토큰으로 accessToken을 재발급 할 수 있기 때문에 외부에서 접근하지 못하는 쿠키에 담았습니다.

**accessToken**
- 사용자의 인가에 사용될 토큰 입니다. accessToken은 쿠키, 응답 body, 헤더 세곳에 저장하는 방법이 많았는데, 어떤 방법이 제일 좋은지 알지 못해서 우선 가장 안전한 쿠키에 저장했습니다.

dye0p committed:
feat: 비밀번호 일치 여부 검증 기능 구현

- 주어진 비밀번호와 현재 비밀번호가 일치하지 않으면 예외가 발생한다.

dye0p committed:
feat: 로그인 요청을 받기 위한 Dto 구현

dye0p committed:
feat: 로그인 요청 수행 컨트롤러, 서비스 뼈대 코드 구현

dye0p committed:
feat: LoginId로 Member를 찾아 검증하는 기능 구현

- Member를 찾지 못하면 예외가 발생하고, 사용자에게 상태코드 400을 응답한다.

dye0p committed:
feat: Member의 비밀번호를 검증하는 기능 구현

- 요청받은 비밀번호와 회원의 비밀번호가 일치하는지 검증하고, 일치하지 않는다면 예외가 발생한다.
- 비밀번호가 일치하지 않는다면 사용자에게 상태코드 400을 응답한다.

dye0p committed:
build: 토큰 발급을 위한 jwt 의존성 추가

dye0p committed:
feat: jwt secret-key 등록

dye0p committed:
feat: jwt를 생성해주는 JwtProvider 객체 생성

dye0p committed:
feat: jwt 토큰 발급 기능 구현

dye0p committed:
test: 토큰 발급 테스트 디스플레이 네이밍 변경

dye0p committed:
feat: 토큰 만료시간 설정 구현

dye0p committed:
feat: 토큰 subject 추출 기능 구현 및 예외처리 구현

dye0p committed:
feat: refresh 토큰 발급 구현

dye0p committed:
feat: refreshToken 엔티티와 repository 구현

dye0p committed:
feat: 발급된 refreshToken 저장 후 accessToken과 함께 반환 구현

dye0p committed:
feat: 로그인의 응답을 세팅해주기 위한 http 응답 객체 준비

- 응답 쿠키에 세팅해야 하기 때문에 HttpServletResponse 객체 필요

dye0p committed:
feat: 로그인이 성공하면 발급된 토큰을 응답 쿠키에 담는 기능 구현

dye0p committed:
test: 로그인 실패 테스트에 쿠키 존재 검증 추가

dye0p commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/CookieProvider.java:
```diff
@@ -0,0 +1,15 @@
+
+public class CookieProvider {
+
+    public static ResponseCookie createCookie(final String name, final String value) {
```
매개변수에 final을 선언해준 이유는 `ResponseCookie.from()` 메서드를 들어가보니까 주입받는 매개변수가 final로 선언되어 있어서, 외부에서도 final인 것을 명시적으로 표현하기 위해서 사용했습니다!

dye0p commented on backend/fittoring/src/main/resources/application.yml:
```diff
@@ -35,3 +35,7 @@ logging:
       file-name-pattern: logs/application.%d{yyyy-MM}.%i.log
       max-history: 2
       max-file-size: 10MB
+jwt:
+  secret-key: ${JWT_SECRET_KEY}
+  access-token-expiration-millis: ${ACCESS_TOKEN_EXPIRATION_MILLIS}
+  refresh-token-expiration-millis: ${REFRESH_TOKEN_EXPIRATION_MILLIS}
```
SecretKey와 만료시간은 .env 파일에 작성하였습니다.

kimdevv replied:
굳!!!!!!!

ppparkta replied:
p1: 노션에 환경변수 업데이트 부탁드려요!

dye0p replied:
노션에 업데이트 했습니다!

dye0p commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/JwtProvider.java:
```diff
@@ -0,0 +1,87 @@
+    public String createToken(Long memberId) {
+        Date now = new Date();
+        Date accessMillis = new Date(now.getTime() + accessExpirationMillis);
+        return buildToken(memberId, now, accessMillis);
```
payload에 아직은 추가할게 없는 것 같아서 우선 member의 식별자만 넣었습니다. 나중에 권한(role) 같은 것들이 추가되면 그때 추가해도 될 것 같습니다.

kimdevv commented on backend/fittoring/build.gradle:
```diff
@@ -37,6 +41,11 @@ dependencies {
     testImplementation("io.rest-assured:rest-assured:5.5.5")
+
+    //jwt
+    implementation "io.jsonwebtoken:jjwt-api:${jjwtVersion}"
+    runtimeOnly "io.jsonwebtoken:jjwt-impl:${jjwtVersion}"
+    runtimeOnly "io.jsonwebtoken:jjwt-jackson:${jjwtVersion}"
```
p3: 버전을 바로 적어주지 않고 `ext`라는 부분으로 분리해낸 이유가 있을까요??

dye0p replied:
특별한 이유는 아닙니다. 버전 관리를 한곳에서 해주기 위함이었습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/AuthService.java:
```diff
@@ -23,12 +30,34 @@ public void register(SignUpRequest request) {
+        return new AuthTokenResponse(accessToken, refreshToken);
+    }
+
     public void validateDuplicateLoginId(String loginId) {
         if (memberRepository.existsByLoginId(loginId)) {
             throw new DuplicateLoginIdException(BusinessErrorMessage.DUPLICATE_LOGIN_ID.getMessage());
         }
     }
```
p3: 이 메서드가 `register()` 메서드 바로 아래에 위치하면 좋을 것 같아요! 이 클래스에서만 사용되는 메서드라면 `private`로 바꿔도 좋을 것 같습니다~

dye0p replied:
메서드 위치는 이동하겠습니다!

> 이 클래스에서만 사용되는 메서드라면 private로 바꿔도 좋을 것 같습니다~

아이디 중복 검증 요청 API에 사용되는 로직이라 private으로 제한하긴 어려울 것 같습니다ㅠ

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/JwtProvider.java:
```diff
@@ -0,0 +1,87 @@
+        this.refreshExpirationMillis = refreshExpirationMillis;
+    }
+
+    public String createToken(Long memberId) {
```
p3: 리프레시 토큰 생성 메서드와 통일성 있게 이 친구도 `createAccessToken()`이라는 이름을 지어주는 것은 어떠신가요

dye0p replied:
좋네요! 변경하겠습니다.

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/JwtProvider.java:
```diff
@@ -0,0 +1,87 @@
+                .getBody().getSubject();
+    }
+
+    private JwtParser getParse() {
+        return Jwts.parserBuilder()
+                .setSigningKey(secretKey)
+                .build();
+    }
```
p3: `JwtParser`을 생성하는 메서드이니 네이밍을 `getJwtParser`이나 `getParser()`로 하는 건 어떨까요?

그리고 문득 생각난 부분인데, `JwtParser`은 매번 생성될 필요 없이 JwtProvider이 생성되는 시점에 한 번만 생성되고 계속 돌려써도 될 것 같은데 인스턴스 변수로 선언하는 건 좀 아쉬우려나요?

ppparkta replied:
조은데요

dye0p replied:
> JwtParser을 생성하는 메서드이니 네이밍을 getJwtParser이나 getParser()로 하는 건 어떨까요?

좋네요! 반영하겠습니다.

> 인스턴스 변수로 선언하는 건 좀 아쉬우려나요?

아니요! 하나도 안 아쉽습니다. 시크릿 키가 계속해서 변경되는 상황만 아니라면 인스턴스 변수화 하는 것이 매번 객체 생성도 필요없고, 좋다고 생각합니다! 반영할게요!

jye326 replied:
> JwtProvider이 생성되는 시점에 한 번만 생성되고 계속 돌려써

좋네요

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/AuthController.java:
```diff
@@ -23,9 +29,23 @@ public ResponseEntity<Void> signUp(@RequestBody @Valid SignUpRequest request) {
         return new ResponseEntity<>(HttpStatus.CREATED);
     }
 
+    @PostMapping("/signin")
```
p3: `signin`을 보고 순간 회원가입이랑 헷갈려버렸습니다... `login`이라는 명칭은 어떨까요? 메서드 이름도 `login`인만큼!

dye0p replied:
그렇군요.. API 연결 부분이 FE와 연관되어있으니 FE와 상의 후에 변경하겠습니다!

jye326 replied:
p3 : 저도 `login` 한번 어필해봅니다.

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/AuthController.java:
```diff
@@ -23,9 +29,23 @@ public ResponseEntity<Void> signUp(@RequestBody @Valid SignUpRequest request) {
+        setCookie(httpResponse, response);
+        return new ResponseEntity<>(HttpStatus.OK);
+    }
+
```
p3: `login` 메서드가 `setCookie` 메서드 바로 위에 위치하면 좋을 것 같습니다!

dye0p replied:
의견을 이해하지 못했습니다..! 다시 설명해주실 수 있나요?

jye326 replied:
p3 : 동감합니다

kimdevv replied:
현재 이 `login` 메서드가 `setCookie` 메서드를 호출하고 있는데 두 메서드가 떨어져 있어서요! private 메서드는 호출하는 public 메서드 바로 아래에 위치하는 게 읽기 편하니 `login` 메서드의 위치를 `setCookie` 메서드 바로 위로 옮겨달라는 의미였습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/AuthController.java:
```diff
@@ -23,9 +29,23 @@ public ResponseEntity<Void> signUp(@RequestBody @Valid SignUpRequest request) {
         return new ResponseEntity<>(HttpStatus.OK);
     }
+
+    private void setCookie(HttpServletResponse httpResponse, AuthTokenResponse response) {
+        ResponseCookie accessTokenCookie = CookieProvider.createCookie(("accessToken"), response.accessToken());
+        ResponseCookie refreshTokenCookie = CookieProvider.createCookie(("refreshToken"), response.refreshToken());
+        httpResponse.addHeader(HttpHeaders.SET_COOKIE, accessTokenCookie.toString());
+        httpResponse.addHeader(HttpHeaders.SET_COOKIE, refreshTokenCookie.toString());
```
p3: 이 메서드에서는 **토큰을 쿠키에 담는 행위**만 하고 있으니, `setCookie`라는 범용적인 이름보다는 `setTokensInCookie`와 같이 뭔가 **토큰을 쿠키에 담는다**는 의미를 내포하는 네이밍을 사용하는 건 어떨까요?

dye0p replied:
좋은 의견 감사합니다! 우선, 아래 @ppparkta 의 의견을 반영해서 해당 로직을 외부의 클래스로 이동시켰습니다. 
분리하고 네이밍을 다시 생각해 봤을 때, 쿠키에 값을 담는다는 의미만 전달하면 될 것 같아서, 범용적인 의미인 wirte라고 네이밍을 변경하였습니다. 추가적인 의견이 있다면 말씀해주세요!

kimdevv replied:
오 조와요

kimdevv reviewed (commented):
고생하셨습니다~ 지금 보니 고려해야될 부분이 굉장히 많네요 ㄷㄷ 수고 많으셨습니다.

아래와 같은 부분들은 다같이 의논해 보면 좋을 것 같아요! 다음 주에 같이 이야기해 봅시다!
1. 토큰의 유효 기간
2. 액세스 토큰 응답 시 어디에 담을 것인지 (헤더 or 쿠키 ?)
3. 액세스 토큰 payload에 무엇을 넣고, 이를 어떻게 파싱할 것인지

ppparkta commented on backend/fittoring/build.gradle:
```diff
@@ -37,6 +41,11 @@ dependencies {
     testImplementation("io.rest-assured:rest-assured:5.5.5")
+
+    //jwt
+    implementation "io.jsonwebtoken:jjwt-api:${jjwtVersion}"
+    runtimeOnly "io.jsonwebtoken:jjwt-impl:${jjwtVersion}"
+    runtimeOnly "io.jsonwebtoken:jjwt-jackson:${jjwtVersion}"
```
p3: jjwt를 사용하신 이유가 있으신가요? 각각의 의존성을 어떤 용도로 불러오신건지도 궁금해요!

ppparkta replied:
어떤건 implementation고 어떤건 runtimeOnly인데 그 이유도요!

dye0p replied:
그것까진 깊게 고려하지 못했습니다.. 

의존성을 검색해보다. 해당 URL에서 제시하는 의존성이 최신 버전인 것 같아 사용하였습니다!
[geeksforgeeks.org/springboot/spring-boot-3-0-jwt-authentication](https://www.geeksforgeeks.org/springboot/spring-boot-3-0-jwt-authentication-with-spring-security-using-mysql-database/)

>  jjwt를 사용하신 이유가 있으신가요? 각각의 의존성을 어떤 용도로 불러오신건지도 궁금해요!

jjwt를 사용한 이유도 다른 라이브러리를 깊게 비교해보진 않았지만, java에 친화적인 라이브러리이고, 각각의 의존성을 모듈화 하여 필요한 컴포넌트만 사용가능 하다는 장점도 있다고 합니다!

> 어떤건 implementation고 어떤건 runtimeOnly인데 그 이유도요!

이 부분도 직접 정한건 아니고, 가이드에 나와있는 것을 참고했습니다.
가이드를 읽어보니, 
- 컴파일 시점과 런타임 시점에서 필요한 의존성이 각각 따로 있는 것이 경량화에 좋다고 합니다! 
- 시점에 따라서 불필요한 의존성을 없애기 위해서라고 합니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/AuthController.java:
```diff
@@ -23,9 +29,23 @@ public ResponseEntity<Void> signUp(@RequestBody @Valid SignUpRequest request) {
         return new ResponseEntity<>(HttpStatus.OK);
     }
+
+    private void setCookie(HttpServletResponse httpResponse, AuthTokenResponse response) {
+        ResponseCookie accessTokenCookie = CookieProvider.createCookie(("accessToken"), response.accessToken());
+        ResponseCookie refreshTokenCookie = CookieProvider.createCookie(("refreshToken"), response.refreshToken());
+        httpResponse.addHeader(HttpHeaders.SET_COOKIE, accessTokenCookie.toString());
+        httpResponse.addHeader(HttpHeaders.SET_COOKIE, refreshTokenCookie.toString());
```
p3: 토큰을 쿠키에 담는 것도 외부에서 재사용할 수 있는 공통적인 로직으로 보여져요! 도메인(AuthTokenResponse)에 결합된 부분만 남기고 나머지 부분은 별개의 클래스로 분리해내는 것 어떠신가요?

dye0p replied:
좋네요! 추가로 현재 컨트롤러의 책임이 "쿠키 추가"까지 두가지 책임을 가지고 있는 것 같네요. 분리하겠습니다.
분리된 클래스는 static 메서드를 가지도록 만들겠습니다!

분리된 객체 예시
``` java
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class CookieWriter {

    public static void write(HttpServletResponse response, AuthTokenResponse tokens) {
        ResponseCookie accessToken = CookieProvider.createCookie("accessToken", tokens.accessToken());
        ResponseCookie refreshToken = CookieProvider.createCookie("refreshToken", tokens.refreshToken());

        response.addHeader(HttpHeaders.SET_COOKIE, accessToken.toString());
        response.addHeader(HttpHeaders.SET_COOKIE, refreshToken.toString());
    }
}
```

ppparkta replied:
조은데요

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Member.java:
```diff
@@ -41,6 +44,10 @@ public Member(String loginId, String gender, String name, String phone, Password
         this(null, loginId, gender, name, phone, password);
     }
 
+    public void matchPassword(String password) {
+        this.password.validateMatches(password);
+    }
+
```
p3: 캡슐화 좋네요

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/AuthController.java:
```diff
@@ -23,9 +29,23 @@ public ResponseEntity<Void> signUp(@RequestBody @Valid SignUpRequest request) {
     }
 
+    @PostMapping("/signin")
+    public ResponseEntity<Void> login(@RequestBody @Valid SignInRequest request, HttpServletResponse httpResponse) {
```
p3: 유효성 검사 안 하는 것 같은데 `@Valid` 제거해도 될거같습니다!

dye0p replied:
아 유효성 검증 넣는다는 걸 깜빡했네요! DTO에 유효성 검증을 넣어줘야 할 것 같습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/JwtProvider.java:
```diff
@@ -0,0 +1,87 @@
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class JwtProvider {
+
+    private final SecretKey secretKey;
+    private final long accessExpirationMillis;
+    private final long refreshExpirationMillis;
+
+    public JwtProvider(
+            @Value("${jwt.secret-key}") String secretKey,
+            @Value("${jwt.access-token-expiration-millis}") long accessExpirationMillis,
+            @Value("${jwt.refresh-token-expiration-millis}") long refreshExpirationMillis
+    ) {
+        this.secretKey = Keys.hmacShaKeyFor(secretKey.getBytes());
+        this.accessExpirationMillis = accessExpirationMillis;
+        this.refreshExpirationMillis = refreshExpirationMillis;
+    }
```
p3: 이 부분 저도 비슷하게 짰다가 좀 아쉬워서 찾아봤는데 `@ConfigurationProperties` 쓰면 깔끔하게 처리 가능할 것 같아요!

https://blog.yevgnenll.me/posts/spring-configuration-properties-fetch-application-properties

dye0p replied:
좋은 것 같네요! 우선 동작에는 문제가 없어보이니, 우선 순위가 높아보이는 부분부터 변경하고 해당 코멘트 반영해보겠습니다. 좋은 자료 감사합니다 : )

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/JwtProvider.java:
```diff
@@ -0,0 +1,87 @@
+    private String getSubject(String token) {
+        return getParse()
+                .parseClaimsJws(token)
+                .getBody().getSubject();
```
p4: 진짜 진짜 사소한건데 여기서 개행 안하신 이유가 있나요?

dye0p replied:
아! 몰랐네요. 컨벤션에 맞추겠습니다!

ppparkta commented on backend/fittoring/src/test/java/fittoring/mentoring/business/model/PasswordTest.java:
```diff
@@ -20,4 +23,30 @@ void from() {
+        //when
+        //then
+        assertThatThrownBy(() -> password.validateMatches(inputPassword))
+                .isInstanceOf(MisMatchPasswordException.class);
```
p3: 메시지도 검증하면 좋겠네요!

dye0p replied:
반영하겠습니다!

ppparkta reviewed (approved):
처리할 게 정말 많았네요 너무 고생많으셨어요! 팀 내에서 왜 웹 토큰 기반의 인증을 사용하게 되었는지도 본문에 있으면 좋을 것 같습니다. 수고하셨어용

dye0p committed:
refactor: validateDuplicationLoginId() 메서드 순서 변경

dye0p committed:
refactor: accessToken 발급 메서드 네이밍 명확하게 변경(creatToken() -> createAccessToken())

dye0p committed:
refactor: getSubject 메서드 체이닝별 개행

dye0p committed:
refactor: JwtParser 인스턴스 변수화

- 매번 JwtParser를 생성해줄 필요없이 객체 초기화 시점에 한번만 생성하여 계속해서 사용하면, 성능상의 이점과, JwtParser를 불변으로 유지할 수 있다.

dye0p committed:
refactor: Jwt Parse 메서드 네이밍 명확하게 변경(getParse -> getJwtParser)

dye0p committed:
refactor: 쿠키에 Token을 담는 로직 별도의 객체로 분리

- 쿠키를 담는 세부적인 로직 까지 컨트롤러의 책임일 필요가 없다고 판단하였음.

dye0p committed:
test: Password 비밀번호 암호화 예외 테스트, 예외 메세지 검증 추가

dye0p committed:
test: JwtProvider 만료되지 않은 토큰 subject 추출 대기시간 제거

- 만료되지 않은 토큰을 테스트해야하기 때문에 대기시간이 필요없다고 판단함.

dye0p committed:
feat: SignInRequest valid 검증 어노테이션 추가

jye326 commented on backend/fittoring/src/test/java/fittoring/integration/mentoring/api/AuthControllerTest.java:
```diff
@@ -87,6 +90,109 @@ void signUp2() {
+        );
+        memberRepository.save(member);
+
+        SignInRequest request = new SignInRequest("invalidLoginId", "password");
```
p3: 존재하지 않는 id로 요청하기 때문에 유효하지 않은 상황인가요? 구체적인 이유가 없어 id 형식 검증에 걸린 상황 등으로 혼동할 수 있을 것 같아요! 저처럼 '이거 왜 통과 안되지?' 생각하지 않게 DisplayName과 id 값을 변경해보면 어떨까요?
ex. 사용자는 존재하지 않는 아이디로..., "notRegisteredId"

jye326 commented on backend/fittoring/src/test/java/fittoring/integration/mentoring/api/AuthControllerTest.java:
```diff
@@ -87,6 +90,109 @@ void signUp2() {
+        );
+    }
+
+    @DisplayName("사용자는 유효하지 않은 비밀번호로 로그인을 할 수 없고, 쿠키에 토큰이 저장되지 않는다.")
+    @Test
+    void login2() {
+        //given
+        Member member = new Member(
+                "loginId",
+                "이름",
+                "남",
+                "010-1234-5678",
+                Password.from("password")
+        );
+        memberRepository.save(member);
+
+        SignInRequest request = new SignInRequest("loginId", "invalidPassword");
```
p3 : 여기도 '유효하지 않은 비밀번호' 보다는 '패스워드 불일치 상황'이 더 구체적인 것 같아요!

jye326 commented on backend/fittoring/src/test/java/fittoring/mentoring/business/model/PasswordTest.java:
```diff
@@ -20,4 +23,30 @@ void from() {
+                .isInstanceOf(MisMatchPasswordException.class);
+    }
+
+    @DisplayName("비밀번호가 일치하면 정상동작 한다.")
```
p4 : `정상동작 한다`... 맞는 말이긴 한데 왜 이렇게 아쉬울까요?
`예외가 발생하지 않는다`는 어떨까요?

jye326 commented on backend/fittoring/src/test/java/fittoring/mentoring/business/service/JwtProviderTest.java:
```diff
@@ -0,0 +1,77 @@
+        Long memberId = 1L;
+        String token = jwtProvider.createToken(memberId);
+
+        Thread.sleep(2000); // 2초 대기
```
p2: 이거 테스트마다 2초씩 잡아먹나요? 테스트 셋업 시 만료 시간을 더 짧게 하고 밀리세컨드 단위만 기다려보면 어떨까요?
저는 한 20 밀리세컨드 까지는 참을 수 있을 것 같아요

jye326 reviewed (commented):
수고하셨습니다~ 역시 꼼꼼하시네요!@!@!@ 간단한 코멘트 몇 개 남겼으니 확인 부탁드려요!

ppparkta commented:
p1: CI 파이프라인 구성할 때 .env를 불러오지 못하는 이슈가 있어서 깃허브의 환경변수를 직접 등록해서 수동 관리하고 있었어요. (actions>environment variables) 
지금 환경변수로 추가해주신 값들은 제가 수동으로 깃허브에 등록해두었는데, 앞으로 환경변수가 추가되면 깃허브에 직접 입력 부탁드립니다. 사전에 알려드리는 것을 깜빡해서 일단 코멘트로 남겨놓고 월요일에 다시 말씀드릴게요 죄송해요!

dye0p committed:
Merge branch 'feature/173-feat-signup' into feature/163-auth-login

dye0p committed:
refactor: 로그인 API 엔드포인트 변경(/signin -> /login)

- signIn은 signUp과 헷갈리는 문제로 /login으로 변경

jye326 reviewed (approved):
nicework!
